message_passing/message_handler_waiting_compensation_cleanup/workflows.py

```diff
@@ -0,0 +1,121 @@
+    my_activity,
+)
+
+
+@workflow.defn
+class MyWorkflow:
+    """
+    This Workflow upholds the following recommended practices:
+
+    1. The main workflow method ensures that all signal and update handlers are
+       finished before a successful return, and on failure, cancellation, and
+       continue-as-new.
+    2. The update handler performs any necessary compensation/cleanup when the
+       workflow is cancelled, fails, or continues-as-new.
+    """
+
+    def __init__(self) -> None:
+        self.workflow_exit_exception: asyncio.Future[BaseException] = asyncio.Future()
+
+    @workflow.run
+    async def run(self, input: WorkflowInput) -> str:
+        try:
+            # 👉 Use this `try...except` style, instead of waiting for message
+            # handlers to finish in a `finally` block. The reason is that other
+            # exception types will cause a Workflow Task failure, in which case
+            # we do *not* want to wait for message handlers to finish.
+            result = await self._run(input)
+            await workflow.wait_condition(workflow.all_handlers_finished)
+            return result
+        except (
+            asyncio.CancelledError,
+            workflow.ContinueAsNewError,
+            exceptions.FailureError,
+        ) as exc:
+            self.workflow_exit_exception.set_result(exc)
+            await workflow.wait_condition(workflow.all_handlers_finished)
+            raise exc
+
+    @workflow.update
+    async def my_update(self, input: UpdateInput) -> str:
+        """
+        This update handler demonstrates how to handle the situation where the
+        main Workflow method exits prematurely. In that case we perform
+        compensation/cleanup, and fail the Update. The Update caller will get a
+        WorkflowUpdateFailedError.
+        """
+        # Coroutines must be wrapped in tasks in order to use workflow.wait.
+        update_task = asyncio.Task(self._my_update())
+        # 👉 Always use `workflow.wait` instead of `asyncio.wait` in Workflow
+        # code: asyncio's version is non-deterministic.
+        first_completed, _ = await workflow.wait(  # type: ignore
+            [update_task, self.workflow_exit_exception],
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+        # 👉 It's possible that the update completed and the workflow exited
+        # prematurely in the same tick of the event loop. If the Update has
+        # completed, return the Update result to the caller, whether or not the
+        # Workflow is exiting.
+        if (
+            update_task in first_completed
+            or input.on_premature_workflow_exit == OnWorkflowExitAction.CONTINUE
+        ):
+            return await update_task
+        else:
+            await self._my_update_compensation_and_cleanup()
+            raise exceptions.ApplicationError(
+                f"The update failed because the workflow run exited: {await self.workflow_exit_exception}"
+            )
```

Comment:
```suggestion
            raise exceptions.ApplicationError(
                "The update failed because the workflow run exited"
            ) from await self.workflow_exit_exception
```

This retains the stack trace and cause chain and such (though if/when changing the logic to use future exception this logic may be a bit clearer)